grudge/models/euler.py

```diff
@@ -76,7 +79,11 @@ class ConservedEulerField:
 
     @property
     def array_context(self):
-        return self.mass.array_context
+        return (
+            self.mass.array_context
+            if isinstance(self.mass, DOFArray)
+            # NOTE: ConservedEulerField also holds the fluxes sometimes...
+            else self.mass[0].array_context)
```

Comment:
I'm not sure what the status quo is on this, where the variables and the flux share the same container. Is it ok to get the array context like this?

Reply:
Well, I definitely do not love the way this looks. :slightly_smiling_face: 

The root cause of the mess, I think, is that we haven't decided the fate of https://github.com/inducer/arraycontext/issues/162.

IMO, this can go one of two ways that are somewhat sane:

- Either: We say "array containers do not know their array context", this problem here goes away, `DOFArray.array_context` disappears, and we have to contend with large-scale interface breakage, because many places rely on fishing array contexts out of `DOFArray`s.
- Or: we double down on array containers knowing their array context, in which case this struct should receive its own stored `array_context` field that is set by `freeze` and `thaw`. This depends on registration with `with_array_context`, which could conceivably be automated by `with_container_arithmetic`. Note that this is all quite inefficient: For each `freeze` and `thaw`, we traverse the whole container hierarchy twice. Once to do the freeze/thaw, and once again to update all the buried array contexts.

I'm not loving our choices here, but I don't think that continuing without picking one of the two is a great option. We'll just get more mess like the above. I'm somewhat leaning towards the first. How do you all feel about this?

cc @majosm @kaushikcfd @MTCam

Reply:
FWIW, for this particular instance, I've pushed a change that switches this to explicit-actx: 587690a7d83ec10630ed9394291d7b69f36ae18e. This generally feels like the right thing to do, IMO.

Reply:
Hm, this seems like a bigger change and I'm more than happy to revert that commit for this PR.

That said, I'm also mostly in favor of the first option, at least for the case where the callee needs an array context and is forced to fish one out of whatever container it received.

However, how is the first option supposed to work with arithmetic? From what I recall that `_cls_has_array_context_attr` thing was introduced so that the arithmetic knows the array types of the context and can broadcast across object arrays and whatnot.

Reply:
> The more we can simplify the logic there, the better.

Hard agree on that! That chunk of code has been black box of spaghetti to me :cry:

Reply:
I'm possibly misunderstanding, but isn't this what `get_container_context_recursively` is for? Does the extra traversal really contribute much time? Seems like it would be negligible compared to all of the other processing going on.

Reply:
It's not that `get_container_context_recursively` is necessarily costly time-wise (I don't know---it might be?). It's that it can fail: All items in an array container can be, more or less, anything, including scalars. As a result, it can always fail. As a result, any code relying on it is not much better than the spaghetti above, at least IMO.

Reply:
> It's not that `get_container_context_recursively` is necessarily costly time-wise (I don't know---it might be?). It's that it can fail: All items in an array container can be, more or less, anything, including scalars. As a result, it can always fail. As a result, any code relying on it is not much better than the spaghetti above, at least IMO.

Failure =  `get_container_context_recursively` returns `None`?  I guess I might be the only fool waving a flag that says he prefers the 2nd option; double down.